Lib/ast.py

```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
         with self.block(extra=self.get_type_comment(node)):
             self.traverse(node.body)
 
+    def _str_literal_helper(
+        self, value, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
+        """Helper for writing string literals, minimising escapes.
+        Returns (possible quote types, string literal to write).
+        """
+        # Escape characters we've been told to escape, backslashes, and
+        # non-printable characters other than \n and \t.
+        escape = {*escape, '\\'}
+        val = "".join(
+            c.encode('unicode_escape').decode('ascii')
+            if c in escape or (not c.isprintable() and c not in '\n\t') else c
+            for c in value
+        )
```

Comment:
This looks like a very long single expression, what about turning this into a function just like this;
```suggestion
        def escape_char(char):
            if char in (*_custom_escapes, "\\") or (
                not char.isprintable() and char not in "\n\t"
            ):
                return char.encode("unicode_escape").decode("ascii")
            else:
                return char
        escaped_string = "".join(map(escape_char, string))
```

Reply:
I think the original is clearer, myself, mostly because it avoids `map()`.

Reply:
Initially, I failed to understand the whole thing since the designated result from the if-expr was in the first line (since it doesn't fit to 79 chars). Just mapping to a single function and giving the returned iterator to something that consumes iterables is something I found reasonable, and IMHO the seperation of the if-exp to its own function is just makes things self-explaning.

Reply:
I parenthesised and indented the ternary expression. I think this strikes a good balance, but happy to change it if you still feel a function is better.